packages/flutter_tools/lib/src/android/deferred_components.dart

```diff
@@ -0,0 +1,674 @@
+          goldenComparisonResults = _checkLoadingUnitGolden(generatedLoadingUnits, deferredComponentsGolden);
+          break;
+        case DeferredComponentsSetupVerificationCheck.writeGolden:
+          _writeGolden(generatedLoadingUnits);
+          break;
+      }
+    }
+    _handleResults(generatedFiles, modifiedFiles, goldenComparisonResults, exitOnFail);
+  }
+
+  // Generates additions to settings.gradle, app/build.gradle
+  List<String> _checkAndroidRootSettings(List<DeferredComponent> components) {
+    Directory androidDir = env.projectDir.childDirectory('android');
+    List<String> modifiedFiles = <String>[];
+
+    // TODO(garyq): Re-work this to be much more robust.
+    // settings.gradle add ':module'
+    File settingsGradle = androidDir.childFile('settings.gradle');
+    File settingsGradleOutput = _outputDir.childFile('settings.gradle');
+    bool componentFound = false;
+    List<String> lines = settingsGradle.readAsLinesSync();
+    if (settingsGradleOutput.existsSync()) settingsGradleOutput.deleteSync();
+    settingsGradleOutput.createSync(recursive: true);
+    for (String line in lines) {
+      if (line.length >= 7 && line.substring(0, 7) == 'include') {
+        List<String> elements = line.substring(7).split(', ');
+        for (DeferredComponent component in components) {
+          for (int i = 1; i < elements.length; i++) {
+            if (elements[i] == '\':${component.name}\'') {
+              componentFound = true;
+              break;
+            }
+          }
+          if (!componentFound) {
+            line += ', \':${component.name}\'';
+          }
+        }
+      }
+      settingsGradleOutput.writeAsStringSync('$line\n', mode: FileMode.append, flush: true);
+    }
+    if (componentFound) {
+      settingsGradleOutput.deleteSync();
+    } else {
+      modifiedFiles.add(settingsGradleOutput.path);
+    }
+
+    // TODO(garyq): Re-work this to be much more robust.
+    // app/build.gradle add deferred components = [':modules', ...]
+    componentFound = false;
+    File appBuildGradle = androidDir.childDirectory('app').childFile('build.gradle');
+    File appBuildGradleOutput = _outputDir.childDirectory('app').childFile('build.gradle');
+    lines = appBuildGradle.readAsLinesSync();
+    if (appBuildGradleOutput.existsSync()) appBuildGradleOutput.deleteSync();
+    appBuildGradleOutput.createSync(recursive: true);
+    bool inAndroidBlock = false;
+    int androidStartLineIndex = 0;
+    int androidEndLineIndex = 0;
+    for (int lineNum = 0; lineNum < lines.length; lineNum++) {
+      String line = lines[lineNum];
+      if (line.contains('android') && line.contains('{')) {
+        inAndroidBlock = true;
+        androidStartLineIndex = lineNum;
+      } else if (inAndroidBlock && line.length > 0 && line.substring(0, 1) == '}') {
+        inAndroidBlock = false;
+        androidEndLineIndex = lineNum;
+        String newLine = '    dynamicFeatures = [';
```

Comment:
@blasten here we're trying to add entries into the `android { }` section of a user's build.gradle. We can just ask them to `apply from: 'deferred_components.gradle'` and then we'll just own that file and freely manipulate it right?

Reply:
I was able to remove the need to add this section manually. It is now added programmatically in flutter.gradle.